### PR TITLE
Issue 5363

### PR DESCRIPTION
**Pull request by annstriganova**

### CODE REVIEW

File: aggregation-log-filter-core/src/main/java/com/axibase/tsd/collector/writer/AtsdWriterFactory.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
+package com.axibase.tsd.collector.writer;
+
+import java.nio.channels.WritableByteChannel;
+
+/**
+ * Created by Anna Striganova on 06.06.18
+ */
+public class AtsdWriterFactory {
+    public static WritableByteChannel getWriter(String url, String scheme, String host, int port) {
```
Comment (raipc) [inline aggregation-log-filter-core/src/main/java/com/axibase/tsd/collector/writer/AtsdWriterFactory.java]: url parameter should be enough for factory.

File: aggregation-log-filter-core/src/main/java/com/axibase/tsd/collector/writer/AtsdWriterFactory.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        WritableByteChannel writer = null;
+        final WriterType writerType = WriterType.valueOf(scheme.toUpperCase());
+        switch (writerType) {
+            case UDP:
```
Comment (raipc) [inline aggregation-log-filter-core/src/main/java/com/axibase/tsd/collector/writer/AtsdWriterFactory.java]: Either delete WriterType and switch over schemes, or implement a method WriterType.createWriter(String host, int port)

File: aggregation-log-filter-log4j/src/main/java/com/axibase/tsd/collector/log4j/Log4jCollector.java [hunk 3]
```diff
@@ -160,65 +160,34 @@ private void initSeriesSenderConfig() {
...
-        try {
-            final WriterType writerType = WriterType.valueOf(scheme.toUpperCase());
-            this.writer = (WritableByteChannel) writerType.getWriterClass().newInstance();
-        } catch (InstantiationException e) {
-            final String msg = "Could not create writer instance by type, "
-                    + e.getMessage();
-            LogLog.error(msg);
-        } catch (Exception e) {
-            LogLog.error("Could not instantiate writerType. " + e.getMessage());
+        if (writerPort <= 0) {
```
Comment (raipc) [inline aggregation-log-filter-log4j/src/main/java/com/axibase/tsd/collector/log4j/Log4jCollector.java]: validate port in Writer implementations

File: aggregation-log-filter-log4j/src/main/java/com/axibase/tsd/collector/log4j/Log4jCollector.java [hunk 4]
```diff
@@ -292,7 +261,7 @@ public void setPattern(String pattern) {
     public void setUrl(String atsdUrl) {
         try {
```
Comment (raipc) [inline aggregation-log-filter-log4j/src/main/java/com/axibase/tsd/collector/log4j/Log4jCollector.java]: Make it a simple setter that sets a value to `url` variable

File: aggregation-log-filter-log4j/src/main/java/com/axibase/tsd/collector/log4j/Log4jMessageWriter.java [hunk 5]
```diff
@@ -35,7 +35,7 @@
 public class Log4jMessageWriter implements MessageWriter<LoggingEvent, String, String> {
     private Map<String, String> tags = new LinkedHashMap<String, String>();
     private String entity = AtsdUtil.resolveHostname();
-    private final Map<Key<String>, CounterWithSum> story = new HashMap<Key<String>, CounterWithSum>();
+    private final Map<Key<String>, CounterWithSum> loggers = new HashMap<Key<String>, CounterWithSum>();
```
Comment (raipc) [inline aggregation-log-filter-log4j/src/main/java/com/axibase/tsd/collector/log4j/Log4jMessageWriter.java]: I think that `loggerEventHistory` is a more appropriate name

File: aggregation-log-filter-log4j/src/main/java/com/axibase/tsd/collector/log4j/Log4jMessageWriter.java [hunk 6]
```diff
@@ -78,7 +78,7 @@ public void writeStatMessages(WritableByteChannel writer,
         long time = System.currentTimeMillis();
 
         // compose & clean
-        for (Iterator<Map.Entry<Key<String>, CounterWithSum>> iterator = story.entrySet().iterator(); iterator.hasNext(); ) {
+        for (Iterator<Map.Entry<Key<String>, CounterWithSum>> iterator = loggers.entrySet().iterator(); iterator.hasNext(); ) {
```
Comment (raipc) [inline aggregation-log-filter-log4j/src/main/java/com/axibase/tsd/collector/log4j/Log4jMessageWriter.java]: Can it be written using simple foreach loop?

File: aggregation-log-filter-logback/src/main/java/com/axibase/tsd/collector/logback/Collector.java [hunk 7]
```diff
@@ -127,7 +129,7 @@ public void start() {
         Map<String, String> stringSettings = new HashMap<>();
         stringSettings.put("debug", debug);
         stringSettings.put("pattern", pattern);
-        stringSettings.put("scheme", scheme);
+        stringSettings.put("url", url);
```
Comment (raipc) [inline aggregation-log-filter-logback/src/main/java/com/axibase/tsd/collector/logback/Collector.java]: Keep existing tags. Simple substring is enough in this case to get scheme.

**Comment by raipc** [comment]
OK, it's a lot better now. But there is a problem if writer is not initialized (e.g. credentials not specified). I think it's a good idea to move `initWriter` logic to `start` method and initialize aggregator andlogbackWriter only if `writer` is not null (factory should throw exception if valid writer cannot be created). Don't forget to check in `stop` method if writer is initialized.

**Comment by rodionos** [comment]
Great work. @raipc - thanks for sharing focused reviews and feedback with Anna.

**Pull request merged**